T6836 VRF module

omnom62 opened the pull request:
<!-- All PR should follow this template to allow a clean and transparent review -->
<!-- Text placed between these delimiters is considered a comment and is not rendered -->

## Change Summary
<!--- Provide a general summary of your changes in the Title above -->

## Types of changes
<!---
What types of changes does your code introduce? Put an 'x' in all the boxes that apply.
NOTE: Markdown requires no leading or trailing whitespace inside the [ ] for checking
the box, please use [x]
-->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)
<!-- optional: Link to related other tasks on Phabricator. -->
<!-- * https://vyos.dev/Txxxx -->

## Related PR(s)
<!-- Link here any PRs in other repositories that are required by this PR -->

## Component(s) name
<!-- A rather incomplete list of components: ethernet, wireguard, bgp, mpls, ldp, l2tp, dhcp ... -->

## Proposed changes
<!--- Describe your changes in detail -->

## How to test
<!---
Please describe in detail how you tested your changes. Include details of your testing
environment, and the tests you ran. When pasting configs, logs, shell output, backtraces,
and other large chunks of text, surround this text with triple backtics
```
like this
```
-->

## Test results
<!--
Provide the output of the unit tests and confirmation of the sanity tests
along with a description of which versions of VyOS you have tested against.

Tests will be run before the PR is accepted, but do not run automatically
on forks, so please run all of the tests with each modification of your PR
to ensure they will pass.

Unit test information can be found in the [Ansible Unit Tests](https://docs.ansible.com/ansible/latest/dev_guide/testing_units.html#testing-units)
section of the documentation.

Sanity test information can be found in the [Ansible Sanity Tests](https://docs.ansible.com/ansible/latest/dev_guide/testing_sanity.html#testing-sanity)
section of the Ansible documentation.

```
Example:

$ ansible-tests units
============================= test session starts ==============================
platform linux -- Python 3.12.2, pytest-8.1.1, pluggy-1.4.0
rootdir: /root/ansible_collections/vyos/vyos
configfile: ../../../ansible/test/lib/ansible_test/_data/pytest/config/default.ini
plugins: xdist-3.5.0, mock-3.14.0
created: 24/24 workers
24 workers [244 items]

........................................................................ [ 29%]
........................................................................ [ 59%]
........................................................................ [ 88%]
............................                                             [100%]
- generated xml file: /root/ansible_collections/vyos/vyos/tests/output/junit/python3.12-controller-units.xml -
============================= 244 passed in 1.55s ==============================

Describe the versions of VyOS that you have tested your changes
against.

```
-->
- [ ] Sanity tests passed
- [ ] Unit tests passed

Tested against VyOS versions:
<!-- examples, add or delete as appropriate; if using rolling versions, please specify
    fully
-->
- 1.3.8
- 1.4-rolling-202201010100


## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!--- The entire development process is outlined here: https://docs.vyos.io/en/latest/contributing/development.html -->
- [ ] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [ ] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the ansible sanity and unit tests
- [ ] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly
- [ ] I have added unit tests to cover my changes
- [ ] I have added a file to `changelogs/fragments` to describe the changes

sever-sever reviewed (commented):
After moving to `FRR-10.5`, (current/stream-2026.02), some `ip protocols` were removed by the upstream (and also removed from our CLI as upstream do  not support it anymore)

This is not valid for the `vyos-stream-2026-02` and the rolling
ref https://github.com/vyos/vyos-1x/pull/4921
```
ip protocol  connected|kernel|table
```
So I propose to remove  ` connected|kernel|table` options from  this implementation

```
vyos_bld@6aa4cba9ed14:/vyos/work/tmp/vyos-ansible-smoketests$ ansible-playbook main.yml 

PLAY [r1] ***********************************************************************************************************************************************************************************

TASK [vrf : Add VRFs] ***********************************************************************************************************************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: vyos@r14#
fatal: [r1]: FAILED! => {"msg": "Unexpected failure during module execution: set vrf name green ip protocol kernel route-map rm1\r\n\r\r\n  \r\n  \r\n  Invalid value\r\n  Value validation failed\r\n  Set failed\r\n\r\n\r[edit]\r\r\nvyos@r14# ", "stdout": ""}

PLAY RECAP **********************************************************************************************************************************************************************************
r1                         : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   

vyos_bld@6aa4cba9ed14:/vyos/work/tmp/vyos-ansible-smoketests$ 
vyos_bld@6aa4cba9ed14:/vyos/work/tmp/vyos-ansible-smoketests$ 
vyos_bld@6aa4cba9ed14:/vyos/work/tmp/vyos-ansible-smoketests$ 
vyos_bld@6aa4cba9ed14:/vyos/work/tmp/vyos-ansible-smoketests$ cat roles/vrf/tasks/main.yml 
---

- name: Add VRFs
  vyos.vyos.vyos_vrf:
    config:
      bind_to_all: true
      instances:
        - name: "green"
          description: "green-vrf"
          table_id: 111
          address_family:
            - afi: "ipv4"
              route_maps:
                - rm_name: "rm1"
                  protocol: "kernel"
                - rm_name: "rm1"
                  protocol: "ospf"

        - name: "red"
          description: "red-vrf"
          table_id: 112


vyos_bld@6aa4cba9ed14:/vyos/work/tmp/vyos-ansible-smoketests$ 

```

Available options:
```
vyos@r14# set vrf name green ip protocol 
Possible completions:
 > any                  Any of the above protocols
 > babel                Babel routing protocol
 > bgp                  Border Gateway Protocol
 > eigrp                Enhanced Interior Gateway Routing Protocol
 > isis                 Intermediate System to Intermediate System
 > ospf                 Open Shortest Path First (OSPFv2)
 > rip                  Routing Information Protocol
 > static               Statically configured routes
 >                      

      
[edit]
vyos@r14#
```

omnom62 commented:
> 1. After moving to `FRR-10.5`, (current/stream-2026.02), some `ip protocols` were removed by the upstream (and also removed from our CLI as upstream do  not support it anymore)
> 
> This is not valid for the `vyos-stream-2026-02` and the rolling ref [vyos/vyos-1x#4921](https://github.com/vyos/vyos-1x/pull/4921)
> 
> ```
> ip protocol  connected|kernel|table
> ```
> 
> So I propose to remove ` connected|kernel|table` options from this implementation
> 
> ```
> vyos_bld@6aa4cba9ed14:/vyos/work/tmp/vyos-ansible-smoketests$ ansible-playbook main.yml 
> 
> PLAY [r1] ***********************************************************************************************************************************************************************************
> 
> TASK [vrf : Add VRFs] ***********************************************************************************************************************************************************************
> An exception occurred during task execution. To see the full traceback, use -vvv. The error was: vyos@r14#
> fatal: [r1]: FAILED! => {"msg": "Unexpected failure during module execution: set vrf name green ip protocol kernel route-map rm1\r\n\r\r\n  \r\n  \r\n  Invalid value\r\n  Value validation failed\r\n  Set failed\r\n\r\n\r[edit]\r\r\nvyos@r14# ", "stdout": ""}
> 
> PLAY RECAP **********************************************************************************************************************************************************************************
> r1                         : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
> 
> vyos_bld@6aa4cba9ed14:/vyos/work/tmp/vyos-ansible-smoketests$ 
> vyos_bld@6aa4cba9ed14:/vyos/work/tmp/vyos-ansible-smoketests$ 
> vyos_bld@6aa4cba9ed14:/vyos/work/tmp/vyos-ansible-smoketests$ 
> vyos_bld@6aa4cba9ed14:/vyos/work/tmp/vyos-ansible-smoketests$ cat roles/vrf/tasks/main.yml 
> ---
> 
> - name: Add VRFs
>   vyos.vyos.vyos_vrf:
>     config:
>       bind_to_all: true
>       instances:
>         - name: "green"
>           description: "green-vrf"
>           table_id: 111
>           address_family:
>             - afi: "ipv4"
>               route_maps:
>                 - rm_name: "rm1"
>                   protocol: "kernel"
>                 - rm_name: "rm1"
>                   protocol: "ospf"
> 
>         - name: "red"
>           description: "red-vrf"
>           table_id: 112
> 
> 
> vyos_bld@6aa4cba9ed14:/vyos/work/tmp/vyos-ansible-smoketests$ 
> ```
> 
> Available options:
> 
> ```
> vyos@r14# set vrf name green ip protocol 
> Possible completions:
>  > any                  Any of the above protocols
>  > babel                Babel routing protocol
>  > bgp                  Border Gateway Protocol
>  > eigrp                Enhanced Interior Gateway Routing Protocol
>  > isis                 Intermediate System to Intermediate System
>  > ospf                 Open Shortest Path First (OSPFv2)
>  > rip                  Routing Information Protocol
>  > static               Statically configured routes
>  >                      
> 
>       
> [edit]
> vyos@r14#
> ```



> 1. After moving to `FRR-10.5`, (current/stream-2026.02), some `ip protocols` were removed by the upstream (and also removed from our CLI as upstream do  not support it anymore)
> 
> This is not valid for the `vyos-stream-2026-02` and the rolling ref [vyos/vyos-1x#4921](https://github.com/vyos/vyos-1x/pull/4921)
> 
> ```
> ip protocol  connected|kernel|table
> ```
> 
> So I propose to remove ` connected|kernel|table` options from this implementation
> 
> ```
> vyos_bld@6aa4cba9ed14:/vyos/work/tmp/vyos-ansible-smoketests$ ansible-playbook main.yml 
> 
> PLAY [r1] ***********************************************************************************************************************************************************************************
> 
> TASK [vrf : Add VRFs] ***********************************************************************************************************************************************************************
> An exception occurred during task execution. To see the full traceback, use -vvv. The error was: vyos@r14#
> fatal: [r1]: FAILED! => {"msg": "Unexpected failure during module execution: set vrf name green ip protocol kernel route-map rm1\r\n\r\r\n  \r\n  \r\n  Invalid value\r\n  Value validation failed\r\n  Set failed\r\n\r\n\r[edit]\r\r\nvyos@r14# ", "stdout": ""}
> 
> PLAY RECAP **********************************************************************************************************************************************************************************
> r1                         : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
> 
> vyos_bld@6aa4cba9ed14:/vyos/work/tmp/vyos-ansible-smoketests$ 
> vyos_bld@6aa4cba9ed14:/vyos/work/tmp/vyos-ansible-smoketests$ 
> vyos_bld@6aa4cba9ed14:/vyos/work/tmp/vyos-ansible-smoketests$ 
> vyos_bld@6aa4cba9ed14:/vyos/work/tmp/vyos-ansible-smoketests$ cat roles/vrf/tasks/main.yml 
> ---
> 
> - name: Add VRFs
>   vyos.vyos.vyos_vrf:
>     config:
>       bind_to_all: true
>       instances:
>         - name: "green"
>           description: "green-vrf"
>           table_id: 111
>           address_family:
>             - afi: "ipv4"
>               route_maps:
>                 - rm_name: "rm1"
>                   protocol: "kernel"
>                 - rm_name: "rm1"
>                   protocol: "ospf"
> 
>         - name: "red"
>           description: "red-vrf"
>           table_id: 112
> 
> 
> vyos_bld@6aa4cba9ed14:/vyos/work/tmp/vyos-ansible-smoketests$ 
> ```
> 
> Available options:
> 
> ```
> vyos@r14# set vrf name green ip protocol 
> Possible completions:
>  > any                  Any of the above protocols
>  > babel                Babel routing protocol
>  > bgp                  Border Gateway Protocol
>  > eigrp                Enhanced Interior Gateway Routing Protocol
>  > isis                 Intermediate System to Intermediate System
>  > ospf                 Open Shortest Path First (OSPFv2)
>  > rip                  Routing Information Protocol
>  > static               Statically configured routes
>  >                      
> 
>       
> [edit]
> vyos@r14#
> ```

@sever-sever thanks!
The current collection v.6.0.0 will also need to support 1.4.x VyOS releases (older 1.5.x and even 1.3.8, but VRF module deliberately skipping it).
We can make a comment in the documentation about the deprecated VyOS features.
i am not sure we need a separate treatment for for the new VyOS versions; and we can mark support of the deprecated features to go in collection v.7.0.0. How does this sound?

sever-sever commented:
> We can make a comment in the documentation about the deprecated VyOS features.

@omnom62  Add a new VRF module and deprecate some features. There is no point in this.
IMHO, just remove 3 protocols (removed FRR zebra filters `connected|kernel|table` ) and the module will fit all versions.
Glad to hear other folks' opinions.

omnom62 commented:
> set vrf name green ip protocol

I dont necessarily disagree but happy to check on others opinion. Right now, there is commitment to support 1.3.8+ up to the recent 1.5.x, and those will likely support the deprecated now features:
`vyos@target142# set vrf name green ip protocol 
Possible completions:
 > any                  Any of the above protocols
 > babel                Babel routing protocol
 > bgp                  Border Gateway Protocol
 > connected            Connected routes (directly attached subnet or host)
 > eigrp                Enhanced Interior Gateway Routing Protocol
 > isis                 Intermediate System to Intermediate System
 > kernel               Kernel routes (not installed via the zebra RIB)
 > ospf                 Open Shortest Path First (OSPFv2)
 > rip                  Routing Information Protocol
 > static               Statically configured routes
 > table                
`
We planned to remove all the deprecated ones in next major collection release (which is 7.0.0)
Obviously, removing is easier for me - no problems here
@gaige @andamasov @dmbaturin  - seeking your opinion here - Thanks
